jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedGraph.java

```diff
@@ -17,111 +17,111 @@
- *
- * @author Lucas J. Scharenbroich
- * @since Sep 10, 2007
  */
 public class AsWeightedGraph<V, E>
-    extends
-    GraphDelegator<V, E>
-    implements
-    Serializable,
-    Graph<V, E>
+    extends GraphDelegator<V, E>
+    implements Serializable, Graph<V, E>
 {
-    private static final long serialVersionUID = 6408608293216853184L;
 
-    private final Map<E, Double> weightMap;
+    private static final long serialVersionUID = -6838132233557L;
+    private final Map<E, Double> weights;
+    private final boolean writeWeightsThrough;
 
     /**
-     * Constructor for AsWeightedGraph.
+     * Constructor for AsWeightedGraph creating a weighted view which does not change the backing
+     * graph's edge weights.
      *
-     * @param g the backing graph over which a weighted view is to be created.
-     * @param weightMap A mapping of edges to weights. If an edge is not present in the weight map,
-     *        the edge weight for the underlying graph is returned. Note that a live reference to
-     *        this map is retained, so if the caller changes the map after construction, the changes
-     *        will affect the AsWeightedGraph instance as well.
+     * @param graph   the backing graph over which an weighted view is to be created.
+     * @param weights the map containing the edge weights.
+     * @throws NullPointerException if the graph or the weights are null.
      */
-    public AsWeightedGraph(Graph<V, E> g, Map<E, Double> weightMap)
+    public AsWeightedGraph(Graph<V, E> graph, Map<E, Double> weights)
     {
-        super(g);
-        this.weightMap = Objects.requireNonNull(weightMap, "Weight map cannot be null");
+        this(graph, weights, false);
     }
 
     /**
-     * {@inheritDoc}
+     * Constructor for AsWeightedGraph.
+     *
+     * @param graph               the backing graph over which an weighted view is to be created
+     * @param weights             the map containing the edge weights
+     * @param writeWeightsThrough if set to true, the weights will get propagated to the backing graph.
+     * @throws NullPointerException     if the graph or the weights are null
+     * @throws IllegalArgumentException if <code>writeWeightsThrough</code> is set to true and
+     *                                  <code>graph</code> is not a weighted graph
      */
-    @Override
-    public void setEdgeWeight(E e, double weight)
+    public AsWeightedGraph(Graph<V, E> graph, Map<E, Double> weights, boolean writeWeightsThrough)
     {
-        if (super.getType().isWeighted()) {
-            super.setEdgeWeight(e, weight);
-        }
+        super(graph);
+        this.weights = Objects.requireNonNull(weights);
+        this.writeWeightsThrough = writeWeightsThrough;
 
-        // Always modify the weight map. It would be a terrible violation
-        // of the use contract to silently ignore changes to the weights.
-        weightMap.put(e, weight);
+        if (this.writeWeightsThrough) {
```

Comment:
Doing this preemptively seems wrong since the weight map is live, which is what we want (i.e. changing an entry in the map post-construction may affect the next call to getEdgeWeight), but changing the weight map directly can't write through.  So I think only setEdgeWeight should write through.

Reply:
I don't agree. If you have a different weights map than the weights of the underlying graph, I would expect them to be changed, if I set the `writeWeightsThrough` true. Otherwise you would end up with inconsistent weights after one weight was changed using the `setEdgeWeight()` method.

Reply:
It's debatable, but the subtext here is that we don't want to introduce incompatibility with existing behavior.